### PR TITLE
[moe] merge moe into main

### LINKED ISSUE #4979: [FEATURE]: support moe
### Describe the feature

support moe and OpenMoE model

### PR DESCRIPTION
**Pull request by oahzxl**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

resolved #4979 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

* update moe module
* support openmoe

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by ver217** [comment]
@oahzxl I notice that you changed code in `colossalai/legacy`. Please run `pytest tests/test_legacy` on your local machine and attach the screenshot here.

File: colossalai/moe/checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+    def save_sharded_optimizer(
+        self,
+        optimizer: Optimizer,
+        checkpoint: Path,
+        gather_dtensor: bool,
+        prefix: str,
+        size_per_shard: int,
+    ):
+        raise NotImplementedError()
```
Comment (ver217) [inline colossalai/moe/checkpoint.py:271]: It seems that all optimizer save/load methods are not impletemented. Does it mean we cannot save optimizer now?

File: colossalai/zero/low_level/low_level_optim.py [hunk 2]
```diff
@@ -265,49 +346,124 @@ def _run_reduction(self):
...
+                    if self._communication_dtype is not None:
+                        flat_grads = flat_grads.to(self._communication_dtype)
 
                 if not self._partition_grads:
-                    dist.all_reduce(flat_grads, group=self.dp_pg)
-                    if flat_grads.dtype != grad_dtype:
-                        flat_grads = flat_grads.to(grad_dtype)
+                    if self.moe_extra_dp_pg is None:
+                        dist.all_reduce(flat_grads, group=self.dp_pg)
+                        if flat_grads.dtype != grad_dtype:
```
Comment (ver217) [inline colossalai/zero/low_level/low_level_optim.py:357]: This function is too complicated. Can you break it into some sub-functions?

File: examples/language/openmoe/benchmark/benchmark_cai_dist.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+set -xue
+
+export NCCL_IB_HCA=mlx5_1:1,mlx5_2:1,mlx5_3:1,mlx5_4:1
```
Comment (ver217) [inline examples/language/openmoe/benchmark/benchmark_cai_dist.sh:5]: IB config should not be added in exmaple

File: examples/language/openmoe/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+colossalai >= 0.1.12
```
Comment (ver217) [inline examples/language/openmoe/requirements.txt:1]: The version is too old. The latese release is `0.3.3`

File: examples/language/openmoe/benchmark/benchmark_cai_dist.sh (hunk 3 above)
Comment (oahzxl) [reply examples/language/openmoe/benchmark/benchmark_cai_dist.sh:5]: done

File: examples/language/openmoe/requirements.txt (hunk 4 above)
Comment (oahzxl) [reply examples/language/openmoe/requirements.txt:1]: done

File: colossalai/zero/low_level/low_level_optim.py (hunk 2 above)
Comment (oahzxl) [reply colossalai/zero/low_level/low_level_optim.py:357]: done

File: colossalai/moe/checkpoint.py (hunk 1 above)
Comment (oahzxl) [reply colossalai/moe/checkpoint.py:271]: yes, it true for now. i forget this part and we update asap

File: examples/language/openmoe/README.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+```bash
+CUDA_EXT=1 pip install -U git+https://github.com/hpcaitech/ColossalAI
+```
+
+Then install dependencies.
+
+```bash
+cd ColossalAI
+pip install -r requirements.txt
```
Comment (ver217) [inline examples/language/openmoe/README.md:18]: If colossalai is installed correctly, this line is useless.

File: examples/language/openmoe/README.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+### 2. Install kernels (Optional)
+
+We have utilized `Triton`, `FlashAttention` and `Apex` kernel for better performance. They are not necessary but we recommend you to install them to fully utilize your hardware.
+```
+# install triton via pip
+pip install triton
+
+# install flash attention via pip
+pip install flash-attn
```
Comment (ver217) [inline examples/language/openmoe/README.md:33]: `flash-attn==2.0.5` may be compatible with torch 1.13.1. The latest flash-attn requires torch>=2.0

**Comment by oahzxl** [comment]
> @oahzxl I notice that you changed code in `colossalai/legacy`. Please run `pytest tests/test_legacy` on your local machine and attach the screenshot here.

image

File: examples/language/openmoe/README.md (hunk 6 above)
Comment (oahzxl) [reply examples/language/openmoe/README.md:33]: ok, fixed

File: examples/language/openmoe/README.md (hunk 5 above)
Comment (oahzxl) [reply examples/language/openmoe/README.md]: ok, fixed

**Review by ver217 — APPROVED** [review]

**Review by CWHer — APPROVED** [review]

**Pull request merged**